src/threshold_decryption.rs

```diff
@@ -128,18 +133,25 @@ impl<N: NodeIdT> ThresholdDecryption<N> {
     ///
     /// If we have collected enough, returns the decrypted message.
     pub fn handle_message(&mut self, sender_id: &N, message: Message) -> Result<Step<N>> {
+        let mut step = Step::default();
         if self.terminated {
-            return Ok(Step::default()); // Don't waste time on redundant shares.
-        }
-        let Message(share) = message;
-        if !self.is_share_valid(sender_id, &share) {
-            let fault_kind = FaultKind::UnverifiedDecryptionShareSender;
-            return Ok(Fault::new(sender_id.clone(), fault_kind).into());
+            return Ok(step); // Don't waste time on redundant shares.
         }
-        if self.shares.insert(sender_id.clone(), share).is_some() {
-            return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleDecryptionShares).into());
-        }
-        self.try_output()
+        step.extend(match message {
+            Message::Share(share) => {
+                if !self.is_share_valid(sender_id, &share) {
+                    let fault_kind = FaultKind::UnverifiedDecryptionShareSender;
+                    return Ok(Fault::new(sender_id.clone(), fault_kind).into());
+                }
+                if self.shares.insert(sender_id.clone(), share).is_some() {
+                    return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleDecryptionShares).into());
+                }
+                Step::default()
+            },
+            Message::Cipher(ciphertext) => self.set_ciphertext(ciphertext)?,
+        });
+        step.extend(self.try_output()?);
+        Ok(step)
```

Comment:
I'm not sure if this is correct -- should I be extending the step like this?

Reply:
In general, extending `step` this way is fine, but we shoudln't introduce `Message::Cipher`, so we won't need this change here.